File: coin-change-ii.py
# Time Complexity: O(n * m) where n is the number of denominations of the coins and m is the target amount
# Space Complexity: (n)

# Code successfully ran on Leetcode

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [[0] * (amount + 1)] * (len(coins) + 1)
        
        for i in range(len(coins) + 1):
            dp[i][0] = 1
            
        for i in range(1, len(coins) + 1):
            for j in range(1, amount + 1):
                if j < coins[i - 1]:
                    dp[i][j] = dp[i - 1][j]
                    
                else:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]]
                    
        return dp[-1][-1]
        

File: Coin_Change_II.java
// TC : O(coins.length * amount)
// SC : O(coins.length * amount)


class Solution {
    public int change(int amount, int[] coins) {
        
        if(coins.length == 0 || coins == null) return 0;
        
        int[][] dp = new int[coins.length + 1][amount + 1];
        
        for(int i = 0; i < coins.length + 1; i++)
            dp[i][0] = 1;
        
        for(int i = 1; i < amount + 1; i++)
            dp[0][i] = 0;
        
        for(int i = 1; i < coins.length + 1; i++) {
            for(int j = 1; j < amount + 1; j++) {
                if(j < coins[i - 1])
                    dp[i][j] = dp[i-1][j];
                else
                    dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]];
            }
        }
        return dp[coins.length][amount];
    }

}
